File: crates/amux-dashboard/static/state/interactions.mjs
import { effectsDue, withDeadline } from './effects.mjs';

export const phases = Object.freeze(['accepted', 'queued', 'sending', 'running', 'waiting', 'blocked', 'applied', 'noop', 'refused', 'failed', 'reconciled', 'unknown']);
export const settled = new Set(['applied', 'noop', 'refused', 'failed', 'reconciled']);
export const phaseLabels = Object.freeze({accepted:'Accepted', queued:'Queued', sending:'Sending', running:'Running', waiting:'Waiting', blocked:'Needs attention', applied:'Completed', noop:'No change', refused:'Not applied', failed:'Failed', reconciled:'Confirmed', unknown:'Outcome unknown'});
const copy = value => JSON.parse(JSON.stringify(value));
const id = () => 'int_' + globalThis.crypto.randomUUID();

// Transport declarations are independent of outbox policy: non-queueable commands
// (uploads, live controls, auth) still owe feedback and a receipt.
export function commandFor(method, pathname) {
  const segments = pathname.split('/').filter(Boolean);
  const domain = segments[1] || 'unknown';
  const primitive = ({sessions:'worker', workers:'worker', board:'board', schedules:'scheduler',
    fs:'filesystem', file:'filesystem', files:'filesystem', upload:'filesystem', uploads:'filesystem',
    groups:'group', tags:'group', memories:'memory', memory:'memory', messages:'message'})[domain] || 'environment';
  const verb = segments.at(-1);
  const specific = domain === 'sessions' && ['send', 'steer', 'start', 'stop'].includes(verb);
  return {id:id(), kind:specific ? (['send','steer'].includes(verb) ? 'message.' : 'worker.') + verb : primitive + '.' + method.toLowerCase(),
    target:{primitive:specific && ['send','steer'].includes(verb) ? 'message' : primitive, id:segments[2] ? decodeURIComponent(segments[2]) : domain}};
}

export function classify(status, body = {}, locallyQueued = false) {
  if (locallyQueued) return {phase:'queued', message:'Queued on this device'};
  if (status >= 400) return {phase:status < 500 ? 'refused' : 'failed', message:body.error || body.message || `Request failed (${status})`};
  if (body.ok === false || body.error || body.ignored_fields?.length) return {phase:'refused', message:body.error || (body.ignored_fields?.length ? 'Not applied: ignored ' + body.ignored_fields.join(', ') : body.message || 'Not applied')};
  if (body.phase && phases.includes(body.phase)) return {phase:body.phase, message:body.message};
  if (body.queued || body.submission === 'deferred') return {phase:'queued', message:'Queued by server'};
  if (status === 202) return {phase:'running', message:'Accepted; awaiting completion'};
  if (body.applied === false || body.deduped === true) return {phase:'noop', message:body.deduped ? 'Already accepted' : 'No change'};
  if (body.applied === true || body.ok === true || body.submitted === true) return {phase:'applied', message:'Completed'};
  return {phase:'unknown', message:'Response received; outcome unconfirmed'};
}

export function createInteractions({storage, now = Date.now, diagnostic = () => {}, max = 200} = {}) {
  const key = 'amux_interactions_v2';
  let receipts = [];
  const listeners = new Set();
  try {
    const saved = JSON.parse(storage?.getItem(key) || '[]');
    if (Array.isArray(saved)) receipts = saved.filter(r => r?.id && phases.includes(r.phase) && r.command?.kind && Array.isArray(r.effects));
    // A reload is not proof that an in-flight command finished.
    receipts = receipts.map(r => ['sending','accepted','running'].includes(r.phase)
      ? {...r, phase:'unknown', measured:false, why_unmeasured:'Page reloaded before completion was observed'} : r);
    receipts = receipts.map(r => r.effect_sync?.phase === 'syncing'
      ? {...r, effect_sync:{...r.effect_sync, phase:'pending', measured:false, next_attempt_at:0}} : r);
  } catch (error) { diagnostic({verdict:'receipt_restore_failed', error:String(error), measured:true, n_considered:1}); }
  function changed(receipt) {
    const unresolved = receipts.filter(r => !settled.has(r.phase));
    const slots = Math.max(0, max - unresolved.length);
    const complete = slots ? receipts.filter(r => settled.has(r.phase)).slice(-slots) : [];
    receipts = [...unresolved, ...complete].sort((a,b) => a.created_at - b.created_at);
    try { storage?.setItem(key, JSON.stringify(receipts)); }
    catch (error) { diagnostic({verdict:'receipt_persistence_failed', error:String(error), measured:true, n_considered:receipts.length}); }
    for (const listener of listeners) listener(copy(receipt));
    return copy(receipt);
  }
  function update(receiptId, patch) {
    const receipt = receipts.find(r => r.id === receiptId);
    if (!receipt) return null;
    if (patch.phase && !phases.includes(patch.phase)) throw new Error('Unknown interaction phase: ' + patch.phase);
    Object.assign(receipt, patch, {updated_at:now()});
    const severity = ['failed','refused'].includes(receipt.phase) ? 'error' : ['unknown','blocked'].includes(receipt.phase) ? 'warning' : settled.has(receipt.phase) ? 'success' : 'info';
    receipt.feedback = {required:true, persistence:['error','warning'].includes(severity) ? 'durable' : 'until-settled', severity, message:phaseLabels[receipt.phase], ...patch.feedback};
    if (['unknown','failed','refused','blocked'].includes(receipt.phase)) diagnostic({kind:'interaction-receipt', verdict:receipt.phase, interaction_id:receipt.id, command_kind:receipt.command.kind, status:receipt.acknowledgement?.status, measured:true, n_considered:1});
    return changed(receipt);
  }
  const api = {
    accept({command, request, origin = {actor:'human', surface:'dashboard'}, feedback, interactionId}) {
      if (interactionId && receipts.some(r => r.id === interactionId)) return api.get(interactionId);
      const receipt = {id:interactionId || id(), command, origin, request, phase:'accepted',
        feedback:{required:true, persistence:'until-settled', severity:'info', message:'Accepted', ...feedback},
        effects:[], acknowledgement:{}, measured:true, n_considered:1, created_at:now(), updated_at:now()};
      receipts.push(receipt);
      return changed(receipt);
    },
    update,
    get: receiptId => { const receipt = receipts.find(r => r.id === receiptId); return receipt ? copy(receipt) : null; },
    recent: (count = 50) => copy(receipts.slice(-Math.max(1, count))),
    subscribe(listener) { listeners.add(listener); return () => listeners.delete(listener); },
    effectStatus(receiptId, patch) {
      const receipt = receipts.find(r => r.id === receiptId);
      if (!receipt) return;
      receipt.effect_sync = {...receipt.effect_sync, ...patch};
      return changed(receipt);
    },
    effect(receiptId, effect) {
      const receipt = receipts.find(r => r.id === receiptId);
      if (!receipt) return;
      if (effect.id && receipt.effects.some(e => e.id === effect.id)) return;
      receipt.effects.push({...effect, id:effect.id || id(), interaction_id:receiptId});
      changed(receipt);
    },
    async acknowledge(receiptId, response) {
      const status = response.status;
      let body = {};
      if (response.headers.get('content-type')?.includes('json')) {
        try { body = await response.clone().json(); }
        catch (_) { return update(receiptId, {phase:'unknown', measured:false, why_unmeasured:'Invalid acknowledgement JSON', acknowledgement:{status}}); }
      }
      if (!body || typeof body !== 'object') body = {};
      const read = ['GET','HEAD'].includes(api.get(receiptId)?.request?.method);
      const result = read && response.ok && !body.error
        ? {phase:'reconciled', message:'Loaded'}
        : classify(status, body, response.headers.get('X-Amux-Outbox') === 'queued');
      const acknowledgement = {status, locally_queued:response.headers.get('X-Amux-Outbox') === 'queued', applied:body.applied, rev:body.rev || body.global_rev, version:body.version, ignored_fields:body.ignored_fields || [], error:body.error, remedy:body.fix || body.remedy || body.how_to_ack};
      // Effects come from authority, never invented from HTTP success or a rev.
      for (const effect of body.effects || []) api.effect(receiptId, effect);
      const measured = body.measured !== false && result.phase !== 'unknown';
      if (measured && api.get(receiptId)?.measured === false) diagnostic({
        verdict:'interaction_measurement_recovered', interaction_id:receiptId, measured:true, n_considered:1});
      return update(receiptId, {phase:result.phase, acknowledgement, measured,
        why_unmeasured:measured ? undefined : body.why_unmeasured || 'Acknowledgement did not confirm the outcome',
        feedback:{message:String(result.message || phaseLabels[result.phase])}});
    },
    expire(age = 120000) {
      for (const receipt of [...receipts]) {
        if (['accepted','sending','running'].includes(receipt.phase) && now() - receipt.updated_at > age) {
          update(receipt.id, {phase:'unknown', measured:false, why_unmeasured:'Completion has not been observed; inspect before retrying'});
        }
      }
    },
  };
  return api;
}

export function createInteractionPoller({interactions, read, reconcile, diagnostic, limit = 8, timeoutMs = 6000, now = Date.now}) {
  let queue = [];
  let flight = false;
  return async function poll() {
    if (flight) return;
    const pending = interactions.recent(Infinity).filter(r => (!settled.has(r.phase) || effectsDue(r, now()))
      && r.command.kind !== 'filesystem.upload' && !['GET','HEAD'].includes(r.request?.method)
      && !(r.phase === 'queued' && r.acknowledgement?.locally_queued));
    const eligible = new Set(pending.map(r => r.id));
    queue = queue.filter(id => eligible.has(id));
    const scheduled = new Set(queue);
    for (const receipt of pending) if (!scheduled.has(receipt.id)) queue.push(receipt.id);
    // Rotate the whole backlog, not just the most recent history page. New
    // arrivals join behind existing work, and a failed read cannot block peers.
    const batch = queue.splice(0, limit);
    queue.push(...batch);
    flight = true;
    try {
      for (const id of batch) {
        try {
          const before = interactions.get(id);
          if (!before) continue;
          if (settled.has(before.phase)) {
            await withDeadline(() => reconcile(id), timeoutMs);
            continue;
          }
          const server = await withDeadline(signal => read(id, signal), timeoutMs);
          const current = interactions.get(id);
          if (!phases.includes(server?.phase)) throw new Error('Invalid interaction status response');
          if (!current || current.phase !== before.phase || current.updated_at !== before.updated_at
            || JSON.stringify(current.acknowledgement) !== JSON.stringify(before.acknowledgement) || settled.has(current.phase)) continue;
          const acknowledgement = server.acknowledgement || current.acknowledgement;
          const changed = server.phase !== current.phase || server.measured !== current.measured
            || server.why_unmeasured !== current.why_unmeasured
            || JSON.stringify(acknowledgement) !== JSON.stringify(current.acknowledgement)
            || Object.entries(server.feedback || {}).some(([key,value]) => current.feedback[key] !== value);
          if (changed) interactions.update(id, {phase:server.phase, measured:server.measured,
            why_unmeasured:server.why_unmeasured, feedback:server.feedback, acknowledgement});
          await withDeadline(() => reconcile(id), timeoutMs);
        } catch (error) {
          diagnostic({verdict:'interaction_status_poll_failed', interaction_id:id,
            error:String(error), measured:true, n_considered:1, pending_count:pending.length});
        }
      }
    } finally { flight = false; }
  };
}

File: crates/amux-server/src/api/interactions.rs
//! Durable receipts over existing domain mutations, independent of presentation.
use super::AppState;
use crate::db::{interactions, SharedStore, WriteOutcome};
use axum::{
    body::{Body, HttpBody},
    extract::{Path, Query, Request, State},
    http::{HeaderValue, StatusCode},
    middleware::Next,
    response::{IntoResponse, Response},
    routing::get,
    Json, Router,
};
use rusqlite::OptionalExtension;
use serde::Deserialize;
use serde_json::{json, Value};

pub fn routes() -> Router<AppState> {
    Router::new()
        .route("/api/interactions/recent", get(recent))
        .route("/api/interactions/{id}", get(one))
        .route("/api/interactions/{id}/effects", get(effects))
        .route("/api/interactions/{id}/why", get(why))
        .route("/api/debug/interactions", get(debug))
        .route("/api/state/summary", get(summary))
}

fn valid_id(value: &str) -> bool {
    !value.is_empty()
        && value.len() <= 128
        && value
            .bytes()
            .all(|b| b.is_ascii_alphanumeric() || b"_.:-".contains(&b))
}

pub fn tracks(method: &str, path: &str) -> bool {
    matches!(method, "POST" | "PUT" | "PATCH" | "DELETE")
        && path.starts_with("/api/")
        && path != "/api/client-debug"
        && !path.starts_with("/api/speedtest")
}

fn primitive(path: &str) -> (&str, &str) {
    let mut parts = path.trim_start_matches('/').split('/').skip(1);
    let domain = parts.next().unwrap_or("environment");
    let kind = match domain {
        "sessions" if path.ends_with("/send") || path.ends_with("/steer") => "message",
        "sessions" | "workers" => "worker",
        "board" => "board",
        "schedules" => "scheduler",
        "files" | "file" | "fs" | "upload" => "filesystem",
        "groups" => "group",
        "memory" | "memories" => "memory",
        "messages" => "message",
        _ => "environment",
    };
    (kind, parts.next().unwrap_or(domain))
}

pub async fn middleware(
    State(store): State<SharedStore>,
    mut req: Request,
    next: Next,
) -> Response {
    let method = req.method().to_string();
    let path = req.uri().path().to_string();
    if !tracks(&method, &path) {
        return next.run(req).await;
    }
    let id =
        match req.headers().get("x-amux-interaction-id") {
            Some(value) => match value.to_str().ok().filter(|v| valid_id(v)) {
                Some(value) => value.to_string(),
                None => return (
                    StatusCode::BAD_REQUEST,
                    Json(
                        json!({"error":"Invalid interaction id", "measured":true,"n_considered":1}),
                    ),
                )
                    .into_response(),
            },
            None => format!("int_{}", ulid::Ulid::new()),
        };
    let actor = super::request_log::caller_from_headers(req.headers());
    let (target_kind, target_id) = primitive(&path);
    let kind = req
        .headers()
        .get("x-amux-command-kind")
        .and_then(|v| v.to_str().ok())
        .filter(|v| valid_id(v))
        .map(str::to_string)
        .unwrap_or_else(|| format!("{target_kind}.{}", method.to_lowercase()));
    let (record_id, record_path, record_method, record_actor, record_kind) = (
        id.clone(),
        path.clone(),
        method.clone(),
        actor.clone(),
        kind.clone(),
    );
    let (target_kind, target_id) = (target_kind.to_string(), target_id.to_string());
    let now = chrono::Utc::now().timestamp_millis();
    let accepted = store.write_async(move |conn| {
        let existing: Option<(String,String,String)> = conn.query_row(
            "SELECT method,path,actor FROM _amux_interactions WHERE id=?1", [&record_id],
            |r| Ok((r.get(0)?,r.get(1)?,r.get(2)?))).optional()?;
        if let Some(existing) = existing {
            if existing != (record_method,record_path,record_actor) {
                return Err(rusqlite::Error::InvalidParameterName("interaction_identity_conflict".into()));
            }
            conn.execute("UPDATE _amux_interactions SET phase='sending', updated_at=?2, attempts=attempts+1 WHERE id=?1", rusqlite::params![record_id, now])?;
        } else {
            conn.execute("INSERT INTO _amux_interactions (id,command_kind,method,path,actor,target_kind,target_id,phase,created_at,updated_at)
                VALUES (?1,?2,?3,?4,?5,?6,?7,'sending',?8,?8)",
                rusqlite::params![record_id,record_kind,record_method,record_path,record_actor,target_kind,target_id,now])?;
        }
        Ok(WriteOutcome{applied:false,events:vec![]})
    }).await;
    if let Err(error) = accepted {
        tracing::warn!(verdict="interaction_accept_failed", interaction_id=%id, %error, "Command not executed: receipt could not be recorded");
        let status = if error.to_string().contains("interaction_identity_conflict") {
            StatusCode::CONFLICT
        } else {
            StatusCode::SERVICE_UNAVAILABLE
        };
        return (status, Json(json!({"error":"Interaction could not be recorded", "interaction_id":id, "measured":true,"n_considered":1}))).into_response();
    }
    req.headers_mut().insert(
        "x-amux-interaction-id",
        HeaderValue::from_str(&id).expect("validated id"),
    );
    let response = interactions::CURRENT.scope(id.clone(), next.run(req)).await;
    let status = response.status().as_u16();
    // Only bounded, already-sized JSON. Never buffer an upload/download stream
    // or consume a response merely to discover it exceeds a telemetry limit.
    let capture = response
        .headers()
        .get("content-type")
        .and_then(|v| v.to_str().ok())
        .is_some_and(|v| v.contains("json"))
        && response
            .body()
            .size_hint()
            .exact()
            .is_some_and(|n| n <= 65536);
    let (mut response, body) = if capture {
        let (parts, body) = response.into_parts();
        match axum::body::to_bytes(body, 65536).await {
            Ok(bytes) => {
                let value = serde_json::from_slice::<Value>(&bytes).unwrap_or(Value::Null);
                (Response::from_parts(parts, Body::from(bytes)), value)
            }
            Err(error) => {
                tracing::warn!(verdict="interaction_ack_read_failed", interaction_id=%id, %error);
                (Response::from_parts(parts, Body::empty()), Value::Null)
            }
        }
    } else {
        (response, Value::Null)
    };
    let phase = if capture && body.is_null() {
        "unknown"
    } else {
        classify(status, &body)
    };
    // Persist only acknowledgement fields, never request payloads or returned
    // configuration/credential values.
    let mut ack = json!({"status":status});
    for key in [
        "applied",
        "rev",
        "global_rev",
        "version",
        "ignored_fields",
        "error",
        "remedy",
        "fix",
        "how_to_ack",
        "queued",
    ] {
        if let Some(value) = body.get(key) {
            ack[key] = value.clone();
        }
    }
    let record_id = id.clone();
    let result = store.write_async(move |conn| {
        conn.execute("UPDATE _amux_interactions SET phase=?2,status=?3,acknowledgement=?4,updated_at=?5 WHERE id=?1",
            rusqlite::params![record_id,phase,status,ack.to_string(),chrono::Utc::now().timestamp_millis()])?;
        Ok(WriteOutcome{applied:false,events:vec![]})
    }).await;
    if let Err(error) = result {
        tracing::warn!(verdict="interaction_completion_unrecorded", interaction_id=%id, %error);
    }
    if matches!(phase, "failed" | "refused" | "unknown") {
        tracing::warn!(verdict="interaction_outcome", interaction_id=%id, command_kind=%kind, %phase, status, "Command outcome needs inspection");
    }
    response.headers_mut().insert(
        "x-amux-interaction-id",
        HeaderValue::from_str(&id).expect("validated id"),
    );
    response.headers_mut().insert(
        "x-amux-command-kind",
        HeaderValue::from_str(&kind).expect("validated command kind"),
    );
    response
}

pub fn classify(status: u16, body: &Value) -> &'static str {
    if status >= 500 {
        return "failed";
    }
    if status >= 400 {
        return "refused";
    }
    if body.get("error").is_some_and(|v| !v.is_null())
        || body["ok"] == false
        || body["ignored_fields"]
            .as_array()
            .is_some_and(|v| !v.is_empty())
    {
        return "refused";
    }
    match body["phase"].as_str() {
        Some("accepted") => return "accepted",
        Some("queued") => return "queued",
        Some("sending") => return "sending",
        Some("running") => return "running",
        Some("waiting") => return "waiting",
        Some("blocked") => return "blocked",
        Some("applied") => return "applied",
        Some("noop") => return "noop",
        Some("failed") => return "failed",
        Some("refused") => return "refused",
        Some("unknown") => return "unknown",
        Some("reconciled") => return "reconciled",
        _ => {}
    }
    if body["queued"] == true || body["submission"] == "deferred" {
        return "queued";
    }
    if status == 202 {
        return "running";
    }
    if body["applied"] == false || body["deduped"] == true {
        return "noop";
    }
    if body["applied"] == true || body["ok"] == true || body["submitted"] == true {
        return "applied";
    }
    "unknown"
}

fn read_receipt(conn: &rusqlite::Connection, id: &str) -> anyhow::Result<Option<Value>> {
    let value = conn.query_row("SELECT command_kind,method,path,actor,target_kind,target_id,phase,status,acknowledgement,created_at,updated_at,attempts,applied_writes,unjournaled_writes FROM _amux_interactions WHERE id=?1", [id], |r| {
        let phase: String = r.get(6)?;
        let updated: i64 = r.get(10)?;
        let stale = matches!(phase.as_str(),"sending"|"running") && chrono::Utc::now().timestamp_millis() - updated > 120000;
        let phase = if stale { "unknown" } else { &phase };
        let ack: String = r.get(8)?;
        Ok(json!({"id":id,"command":{"id":id,"kind":r.get::<_,String>(0)?,"target":{"primitive":r.get::<_,String>(4)?,"id":r.get::<_,String>(5)?}},
            "origin":{"actor":if r.get::<_,String>(3)?.is_empty() {"human"} else {"agent"},"session":r.get::<_,String>(3)?,"surface":"api"},"request":{"method":r.get::<_,String>(1)?,"path":r.get::<_,String>(2)?},
            "phase":phase,"acknowledgement":serde_json::from_str::<Value>(&ack).unwrap_or(json!({})),
            "feedback":{"required":true,"persistence":"durable","severity":if matches!(phase,"failed"|"refused") {"error"} else if matches!(phase,"unknown"|"blocked") {"warning"} else {"info"}},
            "measured":!stale,"n_considered":1,"why_unmeasured":if stale {Some("No completion observed; request may still be running or process restarted")} else {None},
            "created_at":r.get::<_,i64>(9)?,"updated_at":updated,"attempts":r.get::<_,i64>(11)?,"applied_writes":r.get::<_,i64>(12)?,"unjournaled_writes":r.get::<_,i64>(13)?}))
    }).optional()?;
    Ok(value)
}

type ReadError = (StatusCode, Json<Value>);

fn unavailable() -> ReadError {
    (
        StatusCode::SERVICE_UNAVAILABLE,
        Json(
            json!({"measured":false,"n_considered":0,"why_unmeasured":"Interaction store unavailable"}),
        ),
    )
}

async fn read<T: Send + 'static>(
    store: SharedStore,
    f: impl FnOnce(&rusqlite::Connection) -> anyhow::Result<T> + Send + 'static,
) -> Result<Json<T>, ReadError> {
    tokio::task::spawn_blocking(move || {
        let conn = store.read()?;
        f(&conn)
    })
    .await
    .map_err(|_| unavailable())?
    .map(Json)
    .map_err(|error| {
        tracing::warn!(verdict="interaction_read_failed", %error);
        unavailable()
    })
}

async fn one(State(state): State<AppState>, Path(id): Path<String>) -> Result<Response, ReadError> {
    let Json(value) = read(state.store, move |conn| read_receipt(conn, &id)).await?;
    Ok(match value {
        Some(v) => Json(v).into_response(),
        None => (
            StatusCode::NOT_FOUND,
            Json(json!({"measured":true,"n_considered":0,"error":"Interaction not found"})),
        )
            .into_response(),
    })
}

async fn effects(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> Result<Json<Value>, ReadError> {
    read(state.store, move |conn| {
        let mut rows = interactions::effects(conn,&id)?;
        let more = rows.len() > 1000;
        rows.truncate(1000);
        Ok(json!({"measured":true,"n_considered":rows.len(),"interaction_id":id,"effects":rows,"more":more}))
    }).await
}

async fn why(
    State(state): State<AppState>,
    Path(id): Path<String>,
) -> Result<Json<Value>, ReadError> {
    read(state.store, move |conn| {
        let receipt = read_receipt(conn,&id)?;
        let effects = interactions::effects(conn,&id)?;
        let mut stmt = conn.prepare("SELECT id,status,error_body FROM _amux_request_log WHERE req_meta IS NOT NULL AND json_extract(req_meta,'$.interaction_id')=?1 ORDER BY id DESC LIMIT 20")?;
        let logs = stmt.query_map([&id], |r| Ok(json!({"id":r.get::<_,i64>(0)?,"status":r.get::<_,i64>(1)?,"error":r.get::<_,Option<String>>(2)?})))?.collect::<rusqlite::Result<Vec<_>>>()?;
        Ok(json!({"measured":true,"n_considered":usize::from(receipt.is_some()),"receipt":receipt,"effects":effects,"request_log":logs,
            "effect_coverage":"Journaled writes in the request scope; spawned work must explicitly propagate the interaction scope"}))
    }).await
}

#[derive(Deserialize, Default)]
struct Params {
    scope: Option<String>,
    since_rev: Option<u64>,
    since_h: Option<f64>,
}

fn validate_params(params: &Params) -> Result<(), ReadError> {
    if params
        .scope
        .as_deref()
        .is_some_and(|s| !s.starts_with("worker:") || s.len() <= 7)
    {
        tracing::warn!(
            verdict = "interaction_scope_refused",
            "Unsupported interaction scope"
        );
        return Err((
            StatusCode::BAD_REQUEST,
            Json(json!({"measured":false,"n_considered":0,
            "error":"Unsupported interaction scope; expected worker:<name>",
            "why_unmeasured":"Scope validation failed before querying receipts"})),
        ));
    }
    Ok(())
}

fn recent_rows(conn: &rusqlite::Connection, params: &Params) -> anyhow::Result<Vec<Value>> {
    let cutoff = chrono::Utc::now().timestamp_millis()
        - (params.since_h.unwrap_or(24.0).clamp(0.0, 336.0) * 3600000.0) as i64;
    let actor = params
        .scope
        .as_deref()
        .and_then(|s| s.strip_prefix("worker:"))
        .unwrap_or("");
    let mut stmt = conn.prepare("SELECT i.id FROM _amux_interactions i WHERE i.updated_at>=?1 AND (?2='' OR i.actor=?2 OR (i.target_kind IN ('worker','message') AND i.target_id=?2))
        AND (?3=0 OR i.phase NOT IN ('applied','noop','reconciled') OR EXISTS(SELECT 1 FROM _amux_interaction_effects e WHERE e.interaction_id=i.id AND e.rev>?3)) ORDER BY i.updated_at DESC LIMIT 201")?;
    let ids = stmt
        .query_map(
            rusqlite::params![cutoff, actor, params.since_rev.unwrap_or(0)],
            |r| r.get::<_, String>(0),
        )?
        .collect::<rusqlite::Result<Vec<_>>>()?;
    ids.iter()
        .map(|id| read_receipt(conn, id)?.ok_or_else(|| anyhow::anyhow!("Receipt disappeared")))
        .collect()
}

async fn recent(
    State(state): State<AppState>,
    Query(params): Query<Params>,
) -> Result<Json<Value>, ReadError> {
    validate_params(&params)?;
    read(state.store, move |conn| {
        let mut rows = recent_rows(conn, &params)?;
        let more = rows.len() > 200;
        rows.truncate(200);
        Ok(json!({"measured":true,"n_considered":rows.len(),"interactions":rows,"more":more}))
    })
    .await
}

async fn summary(
    State(state): State<AppState>,
    Query(params): Query<Params>,
) -> Result<Json<Value>, ReadError> {
    validate_params(&params)?;
    read(state.store, move |conn| {
        let mut rows = recent_rows(conn,&params)?;
        let more = rows.len()>200; rows.truncate(200);
        let rev: u64 = conn.query_row("SELECT rev FROM _amux_rev WHERE id=1",[],|r|r.get(0))?;
        let mut effects = vec![];
        for row in &rows { effects.extend(interactions::effects(conn,row["id"].as_str().unwrap_or_default())?); }
        let next: Vec<_> = rows.iter().filter(|r| matches!(r["phase"].as_str(),Some("refused"|"failed"|"blocked"|"unknown"))).map(|r|json!({"interaction_id":r["id"],"reason":r["phase"],"why":format!("/api/interactions/{}/why",r["id"].as_str().unwrap_or_default())})).collect();
        Ok(json!({"measured":true,"n_considered":rows.len(),"rev":rev,"recent_interactions":rows,"recent_effects":effects,"next_actions":next,"more":more,
            "health":{"outbox_pending":null,"why_unmeasured":"Device-local outboxes are measured by each browser"}}))
    }).await
}

async fn debug(
    State(state): State<AppState>,
    Query(params): Query<Params>,
) -> Result<Json<Value>, ReadError> {
    validate_params(&params)?;
    read(state.store, move |conn| {
        let rows = recent_rows(conn,&params)?;
        let mut groups = std::collections::BTreeMap::<(String,String),Vec<String>>::new();
        let mut coverage = [0usize;3];
        for row in &rows {
            let id = row["id"].as_str().unwrap_or_default();
            let count: i64 = conn.query_row("SELECT count(*) FROM _amux_interaction_effects WHERE interaction_id=?1",[id],|r|r.get(0))?;
            coverage[(count as usize).min(2)] += 1;
            groups.entry((row["command"]["kind"].as_str().unwrap_or_default().into(),row["phase"].as_str().unwrap_or_default().into())).or_default().push(id.into());
        }
        let groups: Vec<_> = groups.into_iter().map(|((kind,phase),ids)| json!({"kind":kind,"phase":phase,"count":ids.len(),"sample":ids[0]})).collect();
        Ok(json!({"measured":true,"n_considered":rows.len(),"groups":groups,"sampled":rows.len()>200,"effect_coverage":{"zero":coverage[0],"one":coverage[1],"many":coverage[2]}}))
    }).await
}
